### PR TITLE
Fim registry monitorization improvements

### PR DESCRIPTION
**Pull request by Molter73**
|Related issue|
|---|
|#5576|

## Description

This PR groups together all the changes needed to monitor registry keys and values as separate entities, allowing users to get more detailed information on changes that might occur in their Windows registry. As a result, we have modified the registry scan algorithm, added new configuration options and alter the existing agent fim database.

The synchronization mechanism has also been modified in order to properly exchange information of both files and registry entries between agents and their manager.

On the manager side, a new set of sycheck decoders have been created, these will be in charge of handling registry events.

The list of issues this PR groups, along with the associated PR that fix them are as follows:

|Issue|PR|
|---|---|
|[Minor fixes for registry scan](https://github.com/wazuh/wazuh/issues/6451)|#6452|
|[Update analysisd to support FIM's new Windows registries fields](https://github.com/wazuh/wazuh/issues/6230)|#6405|
|[Fim registries minor fixes](https://github.com/wazuh/wazuh/issues/6323)|#6323|
|[Add specific restriction and ignore options for the registry values.](https://github.com/wazuh/wazuh/issues/6292)|#6321|
|[Adapt the db sync mechanism to properly work with the new registry structure](https://github.com/wazuh/wazuh/issues/5874)| #6133|
|[Change registry database to adapt the synchronization mechanism for Windows registries](https://github.com/wazuh/wazuh/issues/6178)| #6199|
|[Configuration options for the Windows registry rework](https://github.com/wazuh/wazuh/issues/6157)| #6169|
|[Let FIM collect permissions, user and group for Windows registry keys](https://github.com/wazuh/wazuh/issues/4223)| #6073|
|[Add all new Windows registry options to the "on demand" configuration](https://github.com/wazuh/wazuh/issues/6127)|#6137|
|[The option registry_ignore doesn't work when modifying subkeys of registry](https://github.com/wazuh/wazuh/issues/5251)|#6047|
|[Let FIM filter Windows Registry values](https://github.com/wazuh/wazuh/issues/4150)|#6057|
|[Feature request: File Integrity Monitoring - add attribute recursion_level for Windows registry](https://github.com/wazuh/wazuh/issues/4615)| #6050|
|[Modify registry scan algorithm](https://github.com/wazuh/wazuh/issues/5864)|#6030|
|[Add report_changes option to registry keys in Windows](https://github.com/wazuh/wazuh/issues/5865)| #5936|
|[Create a database library for registry integrity monitoring](https://github.com/wazuh/wazuh/issues/5867)|#6017|
|[Change actual structure of the syscheck directories](https://github.com/wazuh/wazuh/issues/5905)|#5943|
|[Create data structures and functions to the Windows registries improvement](https://github.com/wazuh/wazuh/issues/5873)|Commited straight to branch `fim-registry-improvements`|

### <windows_registry>
Tags to configure registry monitoring, It already existed before, but new attributes have been added::

|Name|Description|Allowed values|Default value|
|---|---|---|---|
| report_changes | Report registry value changes. Limited to REG_SZ,REG_MULTI_SZ, REG_DWORD, REG_DWORD_BIG_ENDIAN, REG_QWORD value types | Default value: no | Allowed values: yes, no |
| diff_size_limit | Limit the maximum size of the value which will report diff information with ``report_changes`` enabled. Values bigger than this size will not eport diff information. | Default value: 50MB | Allowed values: Any positive number followed by KB/MB/GB |
| check_all | It modifies the value of all attributes with the prefix ``check_``. | Default value: no | Allowed values: yes, no |
| check_sum | Check the MD5, SHA-1 and SHA-256 hashes of the registry. | Default value: no | Allowed values: yes, no |
| check_sha1sum | Check only the SHA-1 hash of the registries. | Default value: no | Allowed values: yes, no |
| check_md5sum | Check only the MD5 hash of the registries. | Default value: no | Allowed values: yes, no |
| check_sha256sum | Check only the SHA-256 hash of the registries. | Default value: no | Allowed values: yes, no |
| check_size | Check the size of the registries. | Default value: no | Allowed values: yes, no |
| check_owner | Check the owner of the registries. | Default value: no | Allowed values: yes, no |
| check_group | Check the group owner of the registries. | Default value: no | Allowed values: yes, no |
| check_perm | Check the permission of the registries. | Default value: no | Allowed values: yes, no |
| check_mtime | Check the modification time of a registry. | Default value: no | Allowed values: yes, no |
| check_type | Check the type of a value. It is used to notify changes in the values of the monitored registry. | Default value: no | Allowed values: yes, no |
| restrict_key | Limit checks to registries containing the entered sregex in the registry name. | Default value: N/A | Allowed values: sregex |
| restrict_value | Limit checks to registry values containing the entered sregex in the value name. | Default value: N/A | Allowed values: sregex |
| recursion_level | Limits the maximum level of recursion allowed. | Default value: no | Allowed values: yes, no |

### <registry_nodiff>
Similar to tag nodiff (used for directories and files). It's a list of values to not compute the diff.
Allowed values: Any registry path, with value_name added. (Example: `HKEY_LOCAL_MACHINE\SOFTWARE\test_key\value_name`)

|Name|Description|Allowed values|Default value|
|---|---|---|---|
| type | This is a simple regex pattern to filter out files so alerts are not generated. | Default value: N/A | Allowed values: sregex |

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X
- [x] Source installation
- [x] Source upgrade
- [x] Review logs syntax and correct language
- [x] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
- Memory tests for Windows
  - [x] Scan-build report
  - [x] Coverity
  - [x] Dr. Memory
- Memory tests for macOS
  - [x] Scan-build report

- [x] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [x] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

**Comment by Molter73** [comment]
Scan-build reported 2 defects under windows compilation, neither of them were introduced by this PR.

[scan-build-winagent.zip](https://github.com/wazuh/wazuh/files/5604366/scan-build-winagent.zip)

Both server and agent scan-build compilations reported no errors.

**Comment by jesusjimsa** [comment]
There is only one bug reported by scan-build in macOS, but it was not introduced in this pull request.
[scan-build-mac.zip](https://github.com/wazuh/wazuh/files/5608653/scan-build-mac.zip)

**Review by vikman90 — CHANGES_REQUESTED** [review]
Team, this is a great change!

We need to update `wazuh-template.json`.

Thanks!

File: src/syscheckd/registry/registry.c [hunk 1]
```diff
@@ -0,0 +1,977 @@
...
+                                 int opts,
+                                 MD5_CTX *md5_ctx,
+                                 SHA_CTX *sha1_ctx,
+                                 SHA256_CTX *sha256_ctx) {
+    if (opts & CHECK_MD5SUM) {
+        MD5_Update(md5_ctx, buffer, (unsigned)length);
+    }
+
+    if (opts & CHECK_SHA1SUM) {
+        SHA1_Update(sha1_ctx, buffer, length);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c:305]: I've changed the type of the length paramter of `fim_registry_update_digests` to be `size_t` and removed the cast from `MD5_Update`, no warnings were issued by the compiler and seems a little more coherent now.

File: src/syscheckd/registry/registry.c [hunk 2]
```diff
@@ -0,0 +1,977 @@
...
+    }
+
+    if (opts & CHECK_SHA256SUM) {
+        SHA256_Update(sha256_ctx, buffer, length);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c:309]: I've changed the type of the length paramter of `fim_registry_update_digests` to be `size_t` and removed the cast from `MD5_Update`, no warnings were issued by the compiler and seems a little more coherent now.

File: src/syscheckd/registry/registry.c [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+                                os_sha1 sha1_output,
+                                os_sha256 sha256_output) {
+    unsigned char md5_digest[16];
+    unsigned char sha1_digest[SHA_DIGEST_LENGTH];
+    unsigned char sha256_digest[SHA256_DIGEST_LENGTH];
+    int n;
+
+    if (opts & CHECK_MD5SUM) {
+        MD5_Final(md5_digest, md5_ctx);
+        for (n = 0; n < 16; n++) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done!

**Review by vikman90 — APPROVED** [review]

**Pull request merged**